ts/util/legend.ts

```diff
@@ -0,0 +1,159 @@
+    const header = document.createElement("div");
+    header.className =
+      "flex flex-row items-center border-b-1 border-grey-200 cursor-move";
+    const title = document.createElement("span");
+    title.className = "font-sarabun font-thin text-green-200 text-24 my-2 mx-4";
+    title.innerHTML = "Legend";
+    const spacer = document.createElement("div");
+    spacer.className = "flex-auto";
+    const icon = makeIcon("xmark");
+    icon.classList.add(
+      "fa-solid",
+      "fa-xmark",
+      "text-grey-200",
+      "hover:text-green-200",
+      "text-24",
+      "m-2",
+      "mx-4",
+      "z-50",
+      "relative",
+      "cursor-pointer"
+    );
+    icon.addEventListener("click", () => this.container.remove());
+    header.appendChild(title);
+    header.appendChild(spacer);
+    header.appendChild(icon);
+    // grid of legend labels
+    const body = document.createElement("div");
+    body.className = "m-4 grid grid-rows-4 grid-flow-col gap-2";
+    for (const qcStatus of Object.values(qcStatuses)) {
+      const labelContainer = document.createElement("div");
+      labelContainer.className =
+        "flex items-center space-x-2 bg-grey-100 rounded-md font-inter font-medium p-2 text-12";
+      const icon = makeIcon(qcStatus.icon);
+      icon.classList.add("text-black");
+      const label = document.createElement("span");
+      label.innerHTML = qcStatus.label;
+      labelContainer.appendChild(icon);
+      labelContainer.appendChild(label);
+      body.appendChild(labelContainer);
+    }
+
+    // centering window, setTimeout is required for offsetHeight and offsetWidth to load
+    // TODO: fix flash introduced by timeout
+    window.setTimeout(() => {
+      this.container.style.top =
+        (window.innerHeight - this.container.offsetHeight) / 2 + "px";
+      this.container.style.left =
+        (window.innerWidth - this.container.offsetWidth) / 2 + "px";
+    }, 0);
+
+    this.container.append(header);
+    this.container.append(body);
+    // add dragging functionality
+    this.makeDraggable(header);
+  }
+
+  private makeDraggable(windowContainerTop: HTMLElement) {
+    //state
+    let diffX = 0,
+      diffY = 0,
+      previousPosX = 0,
+      previousPosY = 0;
+
+    const dragMouseDown = (e: MouseEvent) => {
+      // Prevent any default action on this element (you can remove if you need this element to perform its default action)
+      e.preventDefault();
+      // Get the mouse cursor position and set the initial previous positions to begin
+      previousPosX = e.clientX;
+      previousPosY = e.clientY;
+      // When the mouse is let go, call the closing event
+      document.onmouseup = closeDragElement;
+      // call a function whenever the cursor moves
+      document.onmousemove = elementDrag;
+    };
+
+    const elementDrag = (e: MouseEvent) => {
+      // Prevent any default action on this element (you can remove if you need this element to perform its default action)
```

Comment:
Seems like this was copied from somewhere? Remove the comment that's not relevant here, but add a reference the original source

Reply:
what did you mean by this? Did you want me to cite the code I used?

Reply:
Yes, include a link in comment when copying a significant chunk of code